components/script/dom/event.rs

```diff
@@ -175,132 +265,256 @@ impl Event {
+                }
+                // Step 5.9.8 Otherwise:
+                else {
+                    // Step 5.9.8.1 Set target to parent.
+                    target = parent.clone();
+
+                    // Step 5.9.8.2 If isActivationEvent is true, activationTarget is null, and target has
+                    // activation behavior, then set activationTarget to target.
+                    if is_activation_event && activation_target.is_none() {
+                        if let Some(element) = parent.downcast::<Element>() {
+                            if element.as_maybe_activatable().is_some() {
+                                activation_target = Some(DomRoot::from_ref(element));
+                            }
+                        }
+                    }
+
+                    // Step 5.9.8.3 Append to an event path with event, parent, target, relatedTarget,
+                    // touchTargets, and slot-in-closed-tree.
+                    self.append_to_path(
+                        &parent,
+                        Some(&target),
+                        related_target.as_deref(),
+                        slot_in_closed_tree,
+                    );
+                }
 
-        // Not specified in dispatch spec overtly; this is because
-        // the legacy canceled activation behavior of a checkbox
-        // or radio button needs to know what happened in the
-        // corresponding pre-activation behavior.
-        let mut pre_activation_result: Option<InputActivationState> = None;
+                // Step 5.9.9 If parent is non-null, then set parent to the result of invoking parent’s
+                // get the parent with event
+                if !done {
+                    parent_or_none = parent.get_the_parent(self);
+                }
 
-        // Step 5.12
-        if is_activation_event {
-            if let Some(maybe_checkbox) = activation_target {
-                pre_activation_result = maybe_checkbox.legacy_pre_activation_behavior();
+                // TODO Step 5.9.10 Set slot-in-closed-tree to false.
             }
-        }
 
-        let timeline_window = DomRoot::downcast::<Window>(target.global())
-            .filter(|window| window.need_emit_timeline_marker(TimelineMarkerType::DOMEvent));
+            // Step 5.10 Let clearTargetsStruct be the last struct in event’s path whose shadow-adjusted target
+            // is non-null.
+            // Step 5.11 Let clearTargets be true if clearTargetsStruct’s shadow-adjusted target,
+            // clearTargetsStruct’s relatedTarget, or an EventTarget object in clearTargetsStruct’s
+            // touch target list is a node and its root is a shadow root; otherwise false.
+            // TODO: Handle touch target list
+            clear_targets = self
+                .path
+                .borrow()
+                .iter()
+                .rev()
+                .find(|segment| segment.shadow_adjusted_target.is_none())
```

Comment:
The spec text above says to find the "the last struct in event’s path whose shadow-adjusted target is non-null" but the find call here seems to be reversed as it's finding the first segment that has `segment.shadow_adjusted_target.is_none()`.

Reply:
Good find!